// mirajs/dist/js/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "DCdom", {
    enumerable: true,
    get: function() {
        return _coding.default;
    }
});
Object.defineProperty(exports, "DCutilities", {
    enumerable: true,
    get: function() {
        return _utilitaires.default;
    }
});
Object.defineProperty(exports, "DCRouter", {
    enumerable: true,
    get: function() {
        return _router.default;
    }
});
Object.defineProperty(exports, "DCAnimate", {
    enumerable: true,
    get: function() {
        return _animation.default;
    }
});
Object.defineProperty(exports, "DCToast", {
    enumerable: true,
    get: function() {
        return _toast.default;
    }
});
Object.defineProperty(exports, "DCAlert", {
    enumerable: true,
    get: function() {
        return _alert.default;
    }
});
Object.defineProperty(exports, "DCPopup", {
    enumerable: true,
    get: function() {
        return _popup.default;
    }
});
Object.defineProperty(exports, "DCTruncate", {
    enumerable: true,
    get: function() {
        return _truncate.default;
    }
});
Object.defineProperty(exports, "DCcaroussel", {
    enumerable: true,
    get: function() {
        return _caroussel.default;
    }
});
Object.defineProperty(exports, "InjectLoader", {
    enumerable: true,
    get: function() {
        return _injectLoader.default;
    }
});
Object.defineProperty(exports, "SimpleCarousel", {
    enumerable: true,
    get: function() {
        return _simpleCarousel.default;
    }
});
Object.defineProperty(exports, "SlideAnimations", {
    enumerable: true,
    get: function() {
        return _slideAminations.default;
    }
});
Object.defineProperty(exports, "DCAutocomplete", {
    enumerable: true,
    get: function() {
        return _autocomplete.default;
    }
});
Object.defineProperty(exports, "DCSwipe", {
    enumerable: true,
    get: function() {
        return _swipe.default;
    }
});
Object.defineProperty(exports, "DCSwipeX", {
    enumerable: true,
    get: function() {
        return _swipe.SwipeX;
    }
});
Object.defineProperty(exports, "DCSwipeY", {
    enumerable: true,
    get: function() {
        return _swipe.SwipeY;
    }
});
Object.defineProperty(exports, "DCSidemenu", {
    enumerable: true,
    get: function() {
        return _sidemenu.default;
    }
});
Object.defineProperty(exports, "DCSidemenu2", {
    enumerable: true,
    get: function() {
        return _sidemenu.Sidemenu2;
    }
});
Object.defineProperty(exports, "DCScroll", {
    enumerable: true,
    get: function() {
        return _scroll.default;
    }
});
Object.defineProperty(exports, "DCTabs", {
    enumerable: true,
    get: function() {
        return _tabs.default;
    }
});
Object.defineProperty(exports, "DrawerContent", {
    enumerable: true,
    get: function() {
        return _drawerContent.default;
    }
});
Object.defineProperty(exports, "DCNotify", {
    enumerable: true,
    get: function() {
        return _notification.default;
    }
});
Object.defineProperty(exports, "FloatMenu", {
    enumerable: true,
    get: function() {
        return _floatMenu.default;
    }
});
var _coding = _interopRequireDefault(require("./src/coding"));
var _utilitaires = _interopRequireDefault(require("./src/utilitaires"));
var _router = _interopRequireDefault(require("./src/router"));
var _animation = _interopRequireDefault(require("./src/components/animation"));
var _toast = _interopRequireDefault(require("./src/components/toast"));
var _alert = _interopRequireDefault(require("./src/components/alert"));
var _notification = _interopRequireDefault(require("./src/components/notification"));
var _popup = _interopRequireDefault(require("./src/components/popup"));
var _truncate = _interopRequireDefault(require("./src/components/truncate"));
var _caroussel = _interopRequireDefault(require("./src/components/caroussel"));
var _injectLoader = _interopRequireDefault(require("./src/components/injectLoader"));
var _simpleCarousel = _interopRequireDefault(require("./src/components/simple_carousel"));
var _slideAminations = _interopRequireDefault(require("./src/components/slide_aminations"));
var _autocomplete = _interopRequireDefault(require("./src/components/autocomplete"));
var _swipe = _interopRequireWildcard(require("./src/components/swipe"));
var _sidemenu = _interopRequireWildcard(require("./src/components/sidemenu"));
var _tabs = _interopRequireDefault(require("./src/components/tabs"));
var _scroll = _interopRequireDefault(require("./src/components/scroll"));
var _drawerContent = _interopRequireDefault(require("./src/components/drawerContent"));
var _floatMenu = _interopRequireDefault(require("./src/components/floatMenu"));
function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _interopRequireWildcard(obj) {
    if (obj && obj.__esModule) {
        return obj;
    } else {
        var newObj = {};
        if (obj != null) {
            for(var key in obj){
                if (Object.prototype.hasOwnProperty.call(obj, key)) {
                    var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {};
                    if (desc.get || desc.set) {
                        Object.defineProperty(newObj, key, desc);
                    } else {
                        newObj[key] = obj[key];
                    }
                }
            }
        }
        newObj.default = obj;
        return newObj;
    }
}
