// mirajs/dist/js/src/components/caroussel.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.default = void 0;
var _coding = _interopRequireDefault(require("../coding"));
var _animation = _interopRequireDefault(require("./animation"));
var _utilitaires = _interopRequireDefault(require("../utilitaires"));
function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
/**
 * @param _elem {HTMLElement|string}
 * @param _params? {{auto: boolean}}
 */ var Caroussel = function(_elem, _params) {
    try {
        var _loop = function(i3) {
            _coding.default.css(_coding.default.get(_elem).querySelectorAll('img')[i3], {
                width: __LEFT + "px",
                height: __LEFT + "px",
                left: i3 * __LEFT + "px",
                top: (__dimensions.height - __LEFT) / 2 + "px"
            });
            _coding.default.event({
                name: 'click,mousemove,touchmove',
                object: _coding.default.get(_elem).querySelectorAll('img')[i3],
                callback: function(e) {
                    e.preventDefault();
                }
            });
            _coding.default.get(_elem).querySelectorAll('img')[i3].onclick = function(e) {
                e.preventDefault();
                slider(i3);
            };
            _coding.default.addClass(_coding.default.get(_elem).querySelectorAll('img')[i3], 'img-' + i3);
            __prev_pos[i3] = getLeft(i3);
        };
        _coding.default.event({
            name: 'resize',
            object: window,
            callback: function() {
                _coding.default.caroussel(_elem, _params);
            }
        });
        var __dimensions = {
            width: _coding.default.get(_elem).offsetWidth,
            height: _coding.default.get(_elem).offsetHeight
        }, __NB = 3, __IMGS = _coding.default.get(_elem).querySelectorAll('img').length, __SCALE = 1.5, __LEFT = __dimensions.width / __NB, __TIME = 200, __prev_pos = [];
        console.log(__dimensions);
        var getActive = function() {
            var pos = -1;
            for(var i = 0; i < __IMGS; i++){
                var elem = _coding.default.get(_elem).querySelectorAll('img')[i];
                if (elem.className.indexOf('active') !== -1) pos = i;
            }
            return pos;
        };
        var getLeft = function(_pos) {
            return _coding.default.get(_elem).querySelectorAll('img')[_pos].offsetLeft;
        };
        var isActive = function(_pos) {
            return getActive() === _pos;
        };
        for(var i3 = 0; i3 < __IMGS; i3++)_loop(i3);
        var debAc = getActive() > 0 ? getActive() - 1 : 0, finAc = getActive() > 0 ? getActive() + 2 : 0;
        for(var i1 = debAc; i1 < finAc; i1++){
            _coding.default.css(_coding.default.get(_elem).querySelectorAll('img')[i1], {
                width: __LEFT + "px",
                height: __LEFT + "px",
                left: i1 * __LEFT + "px"
            });
        }
        var slider = function(_pos) {
            if (!isActive(_pos)) {
                var tab = _coding.default.get(_elem).querySelectorAll('img');
                if (getLeft(_pos) < getLeft(getActive())) {
                    for(var i5 = 0; i5 < __IMGS; i5++)_coding.default.css(tab[i5], {
                        zIndex: "-1",
                        opacity: "0"
                    });
                    _coding.default.css(tab[_pos], {
                        zIndex: "2"
                    });
                    (0, _animation).default({
                        type: "left",
                        object: tab[_pos],
                        value: __LEFT,
                        time: __TIME
                    });
                    (0, _animation).default({
                        type: "scale",
                        object: tab[_pos],
                        value: __SCALE,
                        prev: 1,
                        time: __TIME
                    });
                    (0, _animation).default({
                        type: "opacity",
                        object: tab[_pos],
                        value: 1,
                        prev: 0.5,
                        time: __TIME
                    });
                    var prev = _pos === 0 ? __IMGS - 1 : _pos - 1, next = getActive();
                    (0, _animation).default({
                        type: "left",
                        object: tab[next],
                        value: __LEFT * 2,
                        time: __TIME
                    });
                    (0, _animation).default({
                        type: "scale",
                        object: tab[next],
                        value: 1,
                        prev: __SCALE,
                        time: __TIME
                    });
                    (0, _animation).default({
                        type: "opacity",
                        object: tab[next],
                        value: 0.5,
                        prev: 1,
                        time: __TIME
                    });
                    _coding.default.css(tab[next], {
                        zIndex: "1"
                    });
                    _coding.default.css(tab[prev], {
                        left: "0",
                        opacity: "0.5",
                        zIndex: "1"
                    });
                    _utilitaires.default.delay(__TIME / 1000, function() {
                        for(var i = 0; i < __IMGS; i++)_coding.default.removeClass(tab[i], 'active');
                        _coding.default.addClass(tab[_pos], 'active');
                    });
                } else if (getLeft(_pos) > getLeft(getActive())) {
                    for(var i2 = 0; i2 < __IMGS; i2++)_coding.default.css(tab[i2], {
                        zIndex: "-1",
                        opacity: "0"
                    });
                    _coding.default.css(tab[_pos], {
                        zIndex: "2"
                    });
                    (0, _animation).default({
                        type: "left",
                        object: tab[_pos],
                        value: __LEFT,
                        time: __TIME
                    });
                    (0, _animation).default({
                        type: "scale",
                        object: tab[_pos],
                        value: __SCALE,
                        prev: 1,
                        time: __TIME
                    });
                    (0, _animation).default({
                        type: "opacity",
                        object: tab[_pos],
                        value: 1,
                        prev: 0.5,
                        time: __TIME
                    });
                    var prev1 = getActive(), next1 = _pos === __IMGS - 1 ? 0 : _pos + 1;
                    (0, _animation).default({
                        type: "left",
                        object: tab[prev1],
                        value: 0,
                        time: __TIME
                    });
                    (0, _animation).default({
                        type: "scale",
                        object: tab[prev1],
                        value: 1,
                        prev: __SCALE,
                        time: __TIME
                    });
                    (0, _animation).default({
                        type: "opacity",
                        object: tab[prev1],
                        value: 0.5,
                        prev: 1,
                        time: __TIME
                    });
                    _coding.default.css(tab[prev1], {
                        zIndex: "1"
                    });
                    _coding.default.css(tab[next1], {
                        left: __LEFT * 2 + "px",
                        opacity: "0.5",
                        zIndex: "1"
                    });
                    _coding.default.delay(__TIME / 1000, function() {
                        for(var i = 0; i < __IMGS; i++)_coding.default.removeClass(tab[i], 'active');
                        _coding.default.addClass(tab[_pos], 'active');
                    });
                }
            }
        };
        var setter = function(_pos, _left) {
            var tab = _coding.default.get(_elem).querySelectorAll('img');
            if (_left > 0 && getLeft(_pos) < __LEFT * 2) {
                _left = Math.abs(_left);
                var left = getLeft(_pos) + _left, scale = 1 + _left / (__LEFT * 2), opacity = 0.5 + _left / (__LEFT * 2);
                console.log(__LEFT, left, _left, scale, opacity);
                for(var i = 0; i < __IMGS; i++)_coding.default.css(tab[i], {
                    zIndex: "-1",
                    opacity: "0"
                });
                _coding.default.css(tab[_pos], {
                    zIndex: "2",
                    left: left + "px",
                    transform: "scale(" + scale + ")",
                    opacity: opacity + ""
                });
                var prev = _pos === 0 ? __IMGS - 1 : _pos - 1, next = _pos === __IMGS - 1 ? 0 : _pos + 1;
                _coding.default.css(tab[next], {
                    zIndex: "1",
                    left: getLeft(next) + _left + "px",
                    transform: "scale(" + (__SCALE - scale + 1) + ")",
                    opacity: 1.5 - opacity + ""
                });
                _coding.default.css(tab[prev], {
                    left: "0",
                    opacity: "0.5",
                    zIndex: "1"
                });
            } else if (_left < 0 && getLeft(_pos) > 0) {}
        };
        if (_params !== undefined) {
            if (_params.auto === true) {
                var timer = function(_pos) {
                    slider(_pos);
                    var pos = __IMGS - 1 === _pos ? 0 : _pos + 1;
                    setTimeout(timer, 5000, pos);
                };
                timer(getActive());
            }
        }
    } catch (e) {
        console.log(e);
    }
};
var _default = Caroussel;
exports.default = _default;
