// mirajs/dist/js/src/components/fade.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.default = void 0;
var _coding = _interopRequireDefault(require("../coding"));
var _animation = _interopRequireDefault(require("./animation"));
function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
var Fade = {
    In: function(_id) {
        try {
            _coding.default.show(_id);
            (0, _animation).default({
                type: "opacity",
                object: _coding.default.get(_id),
                value: 0.5,
                prev: 0,
                time: 100
            });
        } catch (e) {
            console.log(e);
        }
    },
    out: function(_id) {
        try {
            (0, _animation).default({
                type: "opacity",
                object: _coding.default.get(_id),
                value: 0,
                prev: 0.5,
                time: 100
            });
            setTimeout(function() {
                _coding.default.hide(_id);
            }, 1000);
        } catch (e) {
            console.log(e);
        }
    }
};
var _default = Fade;
exports.default = _default;
