// mirajs/dist/js/src/components/injectLoader.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.default = void 0;
function _classCallCheck(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
        throw new TypeError("Cannot call a class as a function");
    }
}
function _defineProperties(target, props) {
    for(var i = 0; i < props.length; i++){
        var descriptor = props[i];
        descriptor.enumerable = descriptor.enumerable || false;
        descriptor.configurable = true;
        if ("value" in descriptor) descriptor.writable = true;
        Object.defineProperty(target, descriptor.key, descriptor);
    }
}
function _createClass(Constructor, protoProps, staticProps) {
    if (protoProps) _defineProperties(Constructor.prototype, protoProps);
    if (staticProps) _defineProperties(Constructor, staticProps);
    return Constructor;
}
var InjectLoader = /*#__PURE__*/ function() {
    "use strict";
    function InjectLoader(element, options) {
        if (options === void 0) options = {};
        _classCallCheck(this, InjectLoader);
        this.element = element;
        this.options = Object.assign({}, {
            indeterminate: true,
            background: 'white',
            loaderColor: 'blue',
            value: 0
        }, options);
        this.setLoaderColor();
        this.root = this.createDivWithClass('loader_injection__root');
        // Injections.
        if (this.options.indeterminate) this.injectCircularLoaderIndeterminate();
        else this.injectCircularLoaderDeterminate(this.options.value, this.setIndicator(this.options.value));
        // Style
        this.setRootStyle();
        this.element.style.position = 'relative';
        this.element.appendChild(this.root);
    }
    _createClass(InjectLoader, [
        {
            key: "setUpdateValue",
            value: function setUpdateValue(value) {
                if (this.options.indeterminate) return;
                this.value = value;
                this.injectCircularLoaderDeterminate(value, this.setIndicator(value));
            }
        },
        {
            /**
     * Set indicator method.
     * @returns {number}
     */ key: "setIndicator",
            value: function setIndicator(value) {
                return 126.92 - value * 126.92 / 100;
            }
        },
        {
            /**
     * dismissing loader method.
     */ key: "dismiss",
            value: function dismiss() {
                this.root.remove();
            }
        },
        {
            /**
     * @param {String} className
     * @returns {HTMLElement}
     */ key: "createDivWithClass",
            value: function createDivWithClass(className) {
                var div = document.createElement('div');
                div.setAttribute('class', className);
                return div;
            }
        },
        {
            key: "setLoaderColor",
            value: function setLoaderColor() {
                switch(this.options.loaderColor){
                    case 'red':
                        this.loaderColor = "__color-red";
                        break;
                    case 'green':
                        this.loaderColor = "__color-green";
                        break;
                    case 'gray':
                        this.loaderColor = "__color-fade";
                        break;
                    default:
                        this.loaderColor = "__color-blue";
                        break;
                }
            }
        },
        {
            /**
     * Inject circular indeterminate loader method.
     */ key: "injectCircularLoaderIndeterminate",
            value: function injectCircularLoaderIndeterminate() {
                this.root.innerHTML = "<div>\n        <span class=\"dcx-preloader dcx-preloader__circular __indeterminate " + this.loaderColor + " \"\n              aria-valuenow=\"0\" role=\"progressbar\"><svg class=\"dcx-preloader__svg\" viewBox=\"22 22 44 44\"><circle\n            class=\"dcx-preloader__circle __indeterminate\" cx=\"44\" cy=\"44\" r=\"20.2\" fill=\"none\"\n            stroke-width=\"3.6\"></circle></svg></span></div>";
            }
        },
        {
            /**
     * Inject circular determinate loader method.
     */ key: "injectCircularLoaderDeterminate",
            value: function injectCircularLoaderDeterminate(value, indicate) {
                this.root.innerHTML = "<div>\n            <span class=\"dcx-preloader dcx-preloader__circular __determinate " + this.loaderColor + "\" aria-valuenow=\"0\" role=\"progressbar\">\n                <svg class=\"dcx-preloader__svg\" viewBox=\"22 22 44 44\">\n                    <circle  \n                        class=\"dcx-preloader__circle __determinate\" cx=\"44\" cy=\"44\" r=\"20.2\"\n                        fill=\"none\" stroke-width=\"3.6\"\n                        style=\"stroke-dasharray: 126.92; stroke-dashoffset: " + indicate + "px\">\n                    </circle>\n                </svg>\n                <div class=\"dcx-preloader__label\">\n                    <span>" + value + "%</span>\n                </div>\n            </span></div>";
            }
        },
        {
            /**
     * Root style method.
     */ key: "setRootStyle",
            value: function setRootStyle() {
                this.root.style.position = 'absolute';
                this.root.style.background = this.options.background === 'white' ? 'rgba(255,255,255,0.53)' : 'rgba(0,0,0,0.48)';
                this.root.style.top = '0';
                this.root.style.left = '0';
                this.root.style.height = '100%';
                this.root.style.width = '100%';
                this.root.style.display = 'flex';
                this.root.style.justifyContent = 'center';
                this.root.style.alignItems = 'center';
                this.root.style.cursor = 'wait';
            }
        }
    ]);
    return InjectLoader;
}();
exports.default = InjectLoader;
