// mirajs/dist/js/src/components/toast.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.default = void 0;
var _coding = _interopRequireDefault(require("../coding"));
var _animation = _interopRequireDefault(require("./animation"));
var _utilitaires = _interopRequireDefault(require("../utilitaires"));
function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
/**
 * @param _obj {{text: string, position: string, time: number|string, context: string|HTMLElement, ?bg: string, ?radius: number}}
 */ var __toast = {
    nb: 0,
    bottom: 15
};
var Toast = function(_obj) {
    var ret = null;
    var matches = function() {
        return _obj.text !== undefined && _obj.time !== undefined;
    };
    try {
        if (matches()) {
            var elem = _coding.default.new("div"), context = null;
            _coding.default.setClass(elem, "dc-toast");
            _coding.default.setContent(elem, _obj.text);
            if (_obj.context !== undefined) {
                context = _coding.default.get(_obj.context);
                context.appendChild(elem);
            } else _coding.default.get('body').appendChild(elem);
            _coding.default.css(elem, {
                opacity: "0"
            });
            var bottom = __toast.bottom, pos = 15;
            __toast.nb++;
            __toast.bottom += elem.offsetHeight + 10;
            switch(_obj.position){
                case "center":
                    if (context == null) pos = _utilitaires.default.centerLeft(elem.offsetWidth);
                    else pos = (context.offsetWidth - elem.offsetWidth) / 2;
                    break;
                case "right":
                    if (context == null) pos = window.innerWidth - 15;
                    else pos = context.offsetWidth - 15;
                    pos -= elem.offsetWidth;
                    break;
                default:
                    pos = 15;
                    break;
            }
            if (_obj.bg !== undefined) _coding.default.addClass(elem, "bg-" + _obj.bg);
            if (_obj.radius !== undefined) _coding.default.addClass(elem, "radius-" + _obj.radius + "-px");
            _coding.default.css(elem, {
                bottom: bottom + "px",
                left: pos + "px",
                opacity: "1"
            });
            if (context != null) _coding.default.css(elem, {
                position: "absolute"
            });
            setTimeout(function() {
                (0, _animation).default({
                    type: "bottom",
                    object: elem,
                    value: bottom - 15,
                    time: 200,
                    prev: bottom
                });
                (0, _animation).default({
                    type: "opacity",
                    object: elem,
                    value: 0,
                    prev: 1,
                    time: 200
                });
            }, parseFloat(_obj.time));
            setTimeout(function() {
                __toast.nb--;
                __toast.bottom -= elem.offsetHeight + 10;
                _coding.default.remove(elem);
            }, parseFloat(_obj.time) + 1000);
            ret = elem;
        } else console.log('Object param not match');
    } catch (e) {
        console.log(e);
    }
    return ret;
};
var _default = Toast;
exports.default = _default;
