// mirajs/dist/react/banner.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _card = _interopRequireWildcard(require("./card"));
var _propTypes = _interopRequireDefault(require("prop-types"));
function _assertThisInitialized(self) {
    if (self === void 0) {
        throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
    }
    return self;
}
function _classCallCheck(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
        throw new TypeError("Cannot call a class as a function");
    }
}
function _defineProperties(target, props) {
    for(var i = 0; i < props.length; i++){
        var descriptor = props[i];
        descriptor.enumerable = descriptor.enumerable || false;
        descriptor.configurable = true;
        if ("value" in descriptor) descriptor.writable = true;
        Object.defineProperty(target, descriptor.key, descriptor);
    }
}
function _createClass(Constructor, protoProps, staticProps) {
    if (protoProps) _defineProperties(Constructor.prototype, protoProps);
    if (staticProps) _defineProperties(Constructor, staticProps);
    return Constructor;
}
function _extends() {
    _extends = Object.assign || function(target) {
        for(var i = 1; i < arguments.length; i++){
            var source = arguments[i];
            for(var key in source){
                if (Object.prototype.hasOwnProperty.call(source, key)) {
                    target[key] = source[key];
                }
            }
        }
        return target;
    };
    return _extends.apply(this, arguments);
}
function _getPrototypeOf(o) {
    _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) {
        return o.__proto__ || Object.getPrototypeOf(o);
    };
    return _getPrototypeOf(o);
}
function _inherits(subClass, superClass) {
    if (typeof superClass !== "function" && superClass !== null) {
        throw new TypeError("Super expression must either be null or a function");
    }
    subClass.prototype = Object.create(superClass && superClass.prototype, {
        constructor: {
            value: subClass,
            writable: true,
            configurable: true
        }
    });
    if (superClass) _setPrototypeOf(subClass, superClass);
}
function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _interopRequireWildcard(obj) {
    if (obj && obj.__esModule) {
        return obj;
    } else {
        var newObj = {};
        if (obj != null) {
            for(var key in obj){
                if (Object.prototype.hasOwnProperty.call(obj, key)) {
                    var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {};
                    if (desc.get || desc.set) {
                        Object.defineProperty(newObj, key, desc);
                    } else {
                        newObj[key] = obj[key];
                    }
                }
            }
        }
        newObj.default = obj;
        return newObj;
    }
}
function _possibleConstructorReturn(self, call) {
    if (call && (_typeof(call) === "object" || typeof call === "function")) {
        return call;
    }
    return _assertThisInitialized(self);
}
function _setPrototypeOf(o, p) {
    _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) {
        o.__proto__ = p;
        return o;
    };
    return _setPrototypeOf(o, p);
}
var _typeof = function(obj) {
    "@swc/helpers - typeof";
    return obj && typeof Symbol !== "undefined" && obj.constructor === Symbol ? "symbol" : typeof obj;
};
function _isNativeReflectConstruct() {
    if (typeof Reflect === "undefined" || !Reflect.construct) return false;
    if (Reflect.construct.sham) return false;
    if (typeof Proxy === "function") return true;
    try {
        Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function() {}));
        return true;
    } catch (e) {
        return false;
    }
}
function _createSuper(Derived) {
    var hasNativeReflectConstruct = _isNativeReflectConstruct();
    return function _createSuperInternal() {
        var Super = _getPrototypeOf(Derived), result;
        if (hasNativeReflectConstruct) {
            var NewTarget = _getPrototypeOf(this).constructor;
            result = Reflect.construct(Super, arguments, NewTarget);
        } else {
            result = Super.apply(this, arguments);
        }
        return _possibleConstructorReturn(this, result);
    };
}
var Banner = /*#__PURE__*/ function(_Component) {
    "use strict";
    _inherits(Banner, _Component);
    var _super = _createSuper(Banner);
    function Banner(props) {
        _classCallCheck(this, Banner);
        var _this;
        _this = _super.call(this, props);
        _this.state = {
            show: false
        };
        return _this;
    }
    _createClass(Banner, [
        {
            /**
     * Function to set the negative action on the banner.
     * @returns {Element|string}
     */ key: "setNegativeAction",
            value: function setNegativeAction() {
                var _this = this;
                if (this.props.negative === undefined) return '';
                return(/*#__PURE__*/ _react.default.createElement("button", {
                    className: "dcx-banner-btn",
                    onClick: function() {
                        return _this.props.negative.listener();
                    }
                }, this.props.negative.text));
            }
        },
        {
            /**
     * Function to set the neuter action on the banner.
     * @returns {string|Element}
     */ key: "setNeuterAction",
            value: function setNeuterAction() {
                var _this = this;
                if (this.props.neuter === undefined) return '';
                return(/*#__PURE__*/ _react.default.createElement("button", {
                    className: "dcx-banner-btn",
                    onClick: function() {
                        return _this.props.neuter.listener();
                    }
                }, this.props.neuter.text));
            }
        },
        {
            /**
     * Function to set the positive action on the banner.
     * @returns {string|Element}
     */ key: "setPositiveAction",
            value: function setPositiveAction() {
                var _this = this;
                if (this.props.positive === undefined) return '';
                return(/*#__PURE__*/ _react.default.createElement("button", {
                    className: "dcx-banner-btn",
                    onClick: function() {
                        return _this.props.positive.listener();
                    }
                }, this.props.positive.text));
            }
        },
        {
            /**
     * Function o set the image.
     * @returns {string|Element}
     */ key: "setBannerImage",
            value: function setBannerImage() {
                if (this.props.img === undefined) return '';
                return(/*#__PURE__*/ _react.default.createElement(BannerImage, {
                    src: this.props.img
                }));
            }
        },
        {
            /**
     * Render function.
     * @returns {Element}
     */ key: "render",
            value: function render() {
                return(/*#__PURE__*/ _react.default.createElement(_card.default, _extends({}, this.props), /*#__PURE__*/ _react.default.createElement(_card.CardBody, null, /*#__PURE__*/ _react.default.createElement("div", {
                    className: "dcx-banner-container"
                }, /*#__PURE__*/ _react.default.createElement("div", {
                    className: "dcx-banner-body"
                }, this.setBannerImage(), /*#__PURE__*/ _react.default.createElement("div", {
                    className: "dcx-banner-text"
                }, this.props.message)), /*#__PURE__*/ _react.default.createElement("div", {
                    className: "dcx-banner-actions"
                }, this.setNegativeAction(), this.setNeuterAction(), this.setPositiveAction())))));
            }
        }
    ]);
    return Banner;
}(_react.default.Component);
exports.default = Banner;
/**
 * Props types.
 * @type {{positive: object, img: string, negative: object, message: string, neuter: object}}
 */ Banner.propTypes = {
    img: _propTypes.default.string,
    message: _propTypes.default.string,
    negative: _propTypes.default.object,
    neuter: _propTypes.default.object,
    positive: _propTypes.default.object
};
/**
 * Defaults props values.
 * @type {{positive: undefined, img: undefined, negative: undefined, message: undefined, neuter: undefined}}
 */ Banner.defaultProps = {
    img: undefined,
    message: undefined,
    negative: undefined,
    neuter: undefined,
    positive: undefined
};
var BannerImage = function(props) {
    return(/*#__PURE__*/ _react.default.createElement("div", {
        className: "dcx-banner-image"
    }, /*#__PURE__*/ _react.default.createElement("img", {
        src: props.src,
        alt: "Banner"
    })));
};
