// mirajs/dist/react/card.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.default = exports.CardImage = exports.CardFooter = exports.CardBody = exports.CardHeader = exports.Card = void 0;
var _react = _interopRequireDefault(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
function _extends() {
    _extends = Object.assign || function(target) {
        for(var i = 1; i < arguments.length; i++){
            var source = arguments[i];
            for(var key in source){
                if (Object.prototype.hasOwnProperty.call(source, key)) {
                    target[key] = source[key];
                }
            }
        }
        return target;
    };
    return _extends.apply(this, arguments);
}
function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
/**
 * Carder principale, l'element qui permet d'englober tous les éléments qui ferons le cardre.
 * @param props Les propriétés du cardre. (type : [outlined, ''] pour un cardre avec bordure)
 * @returns {HTMLElement}
 * @constructor
 */ var Card = function(props) {
    return(/*#__PURE__*/ _react.default.createElement("div", _extends({}, props, {
        className: "card " + (props.type === 'outlined' ? 'outlined' : '') + " " + props.className
    }), props.children));
};
/**
 * Propriétés du cardre.
 * @type {{type: String}}
 */ Card.propTypes = {
    type: _propTypes.default.string
};
/**
 * Valeurs par defaut des props.
 * @type {{type: string}}
 */ Card.defaultProps = {
    type: 'default'
};
/**
 * Element du cardre qui vas contenir l'en-tête.
 * @param props
 * @returns {HTMLElement}
 * @constructor
 */ var CardHeader = function(props) {
    return(/*#__PURE__*/ _react.default.createElement("div", _extends({}, props, {
        className: "card-header " + props.className
    }), props.children));
};
/**
 * Element du cardre qui contiendra le corps.
 * @param props
 * @returns {HTMLElement}
 * @constructor
 */ var CardBody = function(props) {
    return(/*#__PURE__*/ _react.default.createElement("div", _extends({}, props, {
        className: "card-body " + props.className
    }), props.children));
};
/**
 * Element du cardre qui permet d'indiquer une image.
 * @param props
 * @returns {*}
 * @constructor
 */ var CardImage = function(props) {
    return(/*#__PURE__*/ _react.default.createElement("div", _extends({}, props, {
        className: "card-image " + props.className
    }), props.children));
};
/**
 * Element du cardre qui permet d'indiquer le pied du cardre.
 * @param props
 * @returns {*}
 * @constructor
 */ var CardFooter = function(props) {
    return(/*#__PURE__*/ _react.default.createElement("div", _extends({}, props, {
        className: "card-footer " + props.className
    }), props.children));
};
exports.Card = Card;
exports.CardHeader = CardHeader;
exports.CardBody = CardBody;
exports.CardFooter = CardFooter;
exports.CardImage = CardImage;
var _default = Card;
exports.default = _default;
