// mirajs/dist/react/expander.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.ExpanderItem = exports.Expander = void 0;
var _react = _interopRequireWildcard(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _js = require("../js");
function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _interopRequireWildcard(obj) {
    if (obj && obj.__esModule) {
        return obj;
    } else {
        var newObj = {};
        if (obj != null) {
            for(var key in obj){
                if (Object.prototype.hasOwnProperty.call(obj, key)) {
                    var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {};
                    if (desc.get || desc.set) {
                        Object.defineProperty(newObj, key, desc);
                    } else {
                        newObj[key] = obj[key];
                    }
                }
            }
        }
        newObj.default = obj;
        return newObj;
    }
}
var _OBJ = {
    title: 'Title',
    subList: []
};
var Expander = function(props) {
    var expand = function expand() {
        setExpanded(function(state) {
            return !state;
        });
        if (!expanded) _js.SlideAnimations.slideDown(slide_ref.current, 500);
        else _js.SlideAnimations.slideUp(slide_ref.current, 500);
    };
    var object = props.object, handleClick = props.handleClick, handleEdit = props.handleEdit, handleDelete = props.handleDelete, parentExpanded = props.parentExpanded;
    var slide_ref = (0, _react).useRef(null);
    var ref = (0, _react).useState(false), expanded = ref[0], setExpanded = ref[1];
    (0, _react).useEffect(function() {
        if (!parentExpanded) {
            _js.SlideAnimations.slideUp(slide_ref.current, 500);
            setExpanded(false);
        }
    }, [
        parentExpanded
    ]);
    return(/*#__PURE__*/ _react.default.createElement("div", {
        className: "expander-list__wrapper " + (expanded ? 'expanded' : '')
    }, /*#__PURE__*/ _react.default.createElement(ExpanderItem, {
        visibled: true,
        item: object,
        expand: expand,
        handleDelete: handleDelete,
        handleEdit: handleEdit,
        handleClick: handleClick
    }), /*#__PURE__*/ _react.default.createElement("div", {
        ref: slide_ref,
        className: "expander-list__sub"
    }, object.subList.map(function(item, i) {
        /*#__PURE__*/ return _react.default.createElement(ExpanderItem, {
            key: i,
            visibled: false,
            item: item,
            expanded: expanded,
            expand: function() {},
            handleDelete: handleDelete,
            handleEdit: handleEdit,
            handleClick: handleClick
        });
    }))));
};
exports.Expander = Expander;
Expander.propTypes = {
    object: _propTypes.default.object.isRequired,
    handleClick: _propTypes.default.func,
    handleEdit: _propTypes.default.func,
    handleDelete: _propTypes.default.func,
    parentExpanded: _propTypes.default.bool
};
Expander.defaultProps = {
    object: _OBJ,
    handleClick: function() {},
    handleEdit: undefined,
    handleDelete: undefined,
    parentExpanded: false
};
var ExpanderItem = function(props) {
    var handleClick = props.handleClick, handleDelete = props.handleDelete, handleEdit = props.handleEdit, visibled = props.visibled, item = props.item, expanded = props.expanded, expand = props.expand;
    return(/*#__PURE__*/ _react.default.createElement(_react.default.Fragment, null, !visibled && item.subList.length > 0 ? /*#__PURE__*/ _react.default.createElement(Expander, {
        handleClick: handleClick,
        handleDelete: handleDelete,
        handleEdit: handleEdit,
        parentExpanded: expanded,
        object: item
    }) : /*#__PURE__*/ _react.default.createElement("div", {
        className: "expander-list__item "
    }, item.subList.length > 0 ? /*#__PURE__*/ _react.default.createElement("div", {
        className: "expander-list__item-icon expander-list__itme-state",
        onClick: function() {
            return expand();
        }
    }, /*#__PURE__*/ _react.default.createElement("i", {
        className: "DGfi-keyboard_arrow_down"
    })) : /*#__PURE__*/ _react.default.createElement(_react.default.Fragment, null), /*#__PURE__*/ _react.default.createElement("div", {
        className: "expander-list__item-title",
        onClick: function() {
            expand();
            handleClick(item);
        }
    }, item.title), /*#__PURE__*/ _react.default.createElement("div", {
        className: "expander-list__options"
    }, handleEdit ? /*#__PURE__*/ _react.default.createElement("div", {
        className: "expander-list__item-icon expander-list__itme-option",
        onClick: function() {
            return handleEdit(item);
        }
    }, /*#__PURE__*/ _react.default.createElement("i", {
        className: "DGfi-createmode_editedit"
    })) : /*#__PURE__*/ _react.default.createElement(_react.default.Fragment, null), handleDelete ? /*#__PURE__*/ _react.default.createElement("div", {
        className: "expander-list__item-icon expander-list__itme-option",
        onClick: function() {
            return handleDelete(item);
        }
    }, /*#__PURE__*/ _react.default.createElement("i", {
        className: "DGfi-delete"
    })) : /*#__PURE__*/ _react.default.createElement(_react.default.Fragment, null)))));
};
exports.ExpanderItem = ExpanderItem;
ExpanderItem.propTypes = {
    expand: _propTypes.default.func,
    handleClick: _propTypes.default.func,
    handleEdit: _propTypes.default.func,
    handleDelete: _propTypes.default.func,
    visibled: _propTypes.default.bool,
    expanded: _propTypes.default.bool
};
ExpanderItem.defaultProps = {
    expand: undefined,
    handleClick: function() {},
    handleEdit: undefined,
    handleDelete: undefined,
    visibled: false,
    expanded: false
};
