// mirajs/dist/react/list.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.List = exports.ListItem = void 0;
var _react = _interopRequireDefault(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
function _extends() {
    _extends = Object.assign || function(target) {
        for(var i = 1; i < arguments.length; i++){
            var source = arguments[i];
            for(var key in source){
                if (Object.prototype.hasOwnProperty.call(source, key)) {
                    target[key] = source[key];
                }
            }
        }
        return target;
    };
    return _extends.apply(this, arguments);
}
function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
var List = function(props) {
    var children = props.children, className = props.className;
    className = className !== undefined ? className : "";
    return(/*#__PURE__*/ _react.default.createElement("div", _extends({}, props, {
        className: "dcx-list " + className
    }), children));
};
var ListItem = function(props) {
    var avatar = props.avatar, icon = props.icon, title = props.title, description = props.description, className = props.className, avatarProps = props.avatarProps, iconProps = props.iconProps, meta = props.meta, size = props.size;
    className = className !== undefined ? size !== undefined ? className + " " + size : className : size !== undefined ? size : "";
    return(/*#__PURE__*/ _react.default.createElement("div", _extends({}, props, {
        className: "dcx-item " + className
    }), icon || avatar ? /*#__PURE__*/ _react.default.createElement("span", {
        className: "dcx-item__avatar"
    }, avatar ? /*#__PURE__*/ _react.default.createElement("img", _extends({}, avatarProps, {
        src: avatar
    })) : /*#__PURE__*/ _react.default.createElement("i", _extends({}, iconProps, {
        className: icon
    }))) : /*#__PURE__*/ _react.default.createElement(_react.default.Fragment, null), /*#__PURE__*/ _react.default.createElement("span", {
        className: "dcx-item__title"
    }, title, description !== undefined ? /*#__PURE__*/ _react.default.createElement("span", {
        className: "dcx-item__descript"
    }, description) : /*#__PURE__*/ _react.default.createElement(_react.default.Fragment, null)), meta !== undefined ? /*#__PURE__*/ _react.default.createElement("span", {
        className: "dcx-item__meta"
    }, meta) : /*#__PURE__*/ _react.default.createElement(_react.default.Fragment, null)));
};
ListItem.propTypes = _extends({}, new _react.default.Component().propTypes, {
    avatar: _propTypes.default.string,
    icon: _propTypes.default.string,
    title: _propTypes.default.string.isRequired,
    descripiton: _propTypes.default.string,
    className: _propTypes.default.string,
    avatarProps: _propTypes.default.object,
    iconProps: _propTypes.default.object,
    meta: _propTypes.default.string,
    size: _propTypes.default.oneOf([
        "small",
        "medium",
        "big"
    ])
});
ListItem.defaultProps = {
    title: "List Item",
    icon: "DGfi-settings"
};
exports.ListItem = ListItem;
exports.List = List;
