// mirajs/dist/react/m_sidemenu.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _js = require("../js");
var _list = require("./list");
function _assertThisInitialized(self) {
    if (self === void 0) {
        throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
    }
    return self;
}
function _classCallCheck(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
        throw new TypeError("Cannot call a class as a function");
    }
}
function _defineProperties(target, props) {
    for(var i = 0; i < props.length; i++){
        var descriptor = props[i];
        descriptor.enumerable = descriptor.enumerable || false;
        descriptor.configurable = true;
        if ("value" in descriptor) descriptor.writable = true;
        Object.defineProperty(target, descriptor.key, descriptor);
    }
}
function _createClass(Constructor, protoProps, staticProps) {
    if (protoProps) _defineProperties(Constructor.prototype, protoProps);
    if (staticProps) _defineProperties(Constructor, staticProps);
    return Constructor;
}
function _getPrototypeOf(o) {
    _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) {
        return o.__proto__ || Object.getPrototypeOf(o);
    };
    return _getPrototypeOf(o);
}
function _inherits(subClass, superClass) {
    if (typeof superClass !== "function" && superClass !== null) {
        throw new TypeError("Super expression must either be null or a function");
    }
    subClass.prototype = Object.create(superClass && superClass.prototype, {
        constructor: {
            value: subClass,
            writable: true,
            configurable: true
        }
    });
    if (superClass) _setPrototypeOf(subClass, superClass);
}
function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _possibleConstructorReturn(self, call) {
    if (call && (_typeof(call) === "object" || typeof call === "function")) {
        return call;
    }
    return _assertThisInitialized(self);
}
function _setPrototypeOf(o, p) {
    _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) {
        o.__proto__ = p;
        return o;
    };
    return _setPrototypeOf(o, p);
}
var _typeof = function(obj) {
    "@swc/helpers - typeof";
    return obj && typeof Symbol !== "undefined" && obj.constructor === Symbol ? "symbol" : typeof obj;
};
function _isNativeReflectConstruct() {
    if (typeof Reflect === "undefined" || !Reflect.construct) return false;
    if (Reflect.construct.sham) return false;
    if (typeof Proxy === "function") return true;
    try {
        Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function() {}));
        return true;
    } catch (e) {
        return false;
    }
}
function _createSuper(Derived) {
    var hasNativeReflectConstruct = _isNativeReflectConstruct();
    return function _createSuperInternal() {
        var Super = _getPrototypeOf(Derived), result;
        if (hasNativeReflectConstruct) {
            var NewTarget = _getPrototypeOf(this).constructor;
            result = Reflect.construct(Super, arguments, NewTarget);
        } else {
            result = Super.apply(this, arguments);
        }
        return _possibleConstructorReturn(this, result);
    };
}
var MSidemenu = /*#__PURE__*/ function(_Component) {
    "use strict";
    _inherits(MSidemenu, _Component);
    var _super = _createSuper(MSidemenu);
    function MSidemenu(props) {
        _classCallCheck(this, MSidemenu);
        var _this;
        _this = _super.call(this, props);
        _this.state = {
            reduced: false,
            open: false
        };
        return _this;
    }
    _createClass(MSidemenu, [
        {
            key: "componentDidMount",
            value: function componentDidMount() {
                var _this = this;
                this.setReduced();
                window.addEventListener("resize", function() {
                    if (!_this.state.reduced) _this.setReduced();
                }, false);
            }
        },
        {
            key: "setReduced",
            value: function setReduced() {
                var innerWidth = window.innerWidth, reduced = false;
                if (innerWidth < 1000 && innerWidth > 800) reduced = true;
                this.setState({
                    reduced: reduced
                });
            }
        },
        {
            key: "reduce",
            value: function reduce(_article_id) {
                var _this = this;
                this.setState({
                    reduced: !this.state.reduced
                }, function() {
                    if (_this.state.reduced) _js.DCdom.addClass("#" + _article_id, "__reduced-sidemenu");
                    else _js.DCdom.removeClass("#" + _article_id, "__reduced-sidemenu");
                });
            }
        },
        {
            key: "open",
            value: function open() {
                this.setState({
                    open: true
                });
            }
        },
        {
            key: "close",
            value: function close() {
                this.setState({
                    open: false
                });
            }
        },
        {
            key: "render",
            value: function render() {
                var _this = this;
                var menu = this.props.menu;
                return(/*#__PURE__*/ _react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/ _react.default.createElement("div", {
                    className: "m_sidemenu " + (this.state.reduced ? "__reduced" : "") + " " + (this.state.open ? "__open" : "__close")
                }, /*#__PURE__*/ _react.default.createElement("div", {
                    className: "m_sidemenu__container"
                }, menu.map(function(item, index) {
                    /*#__PURE__*/ return _react.default.createElement("div", {
                        key: index,
                        className: "m_sidemenu__item " + (item.active ? "__activated" : "") + " " + (item.bottom ? "__bottom" : ""),
                        onClick: function() {
                            if (item.onClick) item.onClick();
                        }
                    }, /*#__PURE__*/ _react.default.createElement("div", {
                        className: "__div_icon"
                    }, item.img ? /*#__PURE__*/ _react.default.createElement("img", {
                        src: item.img,
                        alt: "Item icon"
                    }) : /*#__PURE__*/ _react.default.createElement("i", {
                        className: item.icon
                    })), /*#__PURE__*/ _react.default.createElement("div", {
                        className: "__div_label"
                    }, item.subtitle ? /*#__PURE__*/ _react.default.createElement(_list.ListItem, {
                        icon: null,
                        title: item.title,
                        description: item.subtitle
                    }) : item.title));
                }))), /*#__PURE__*/ _react.default.createElement("div", {
                    className: "m_fade " + (this.state.open ? "__open" : ""),
                    onClick: function() {
                        return _this.close();
                    }
                })));
            }
        }
    ]);
    return MSidemenu;
}(_react.default.Component);
MSidemenu.propTypes = {
    menu: _propTypes.default.array
};
var _default = MSidemenu;
exports.default = _default;
