// mirajs/dist/react/options_popup.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.OptionsPopup = void 0;
var _react = _interopRequireDefault(require("react"));
var _jsManagement = require("../js/src/components/js_management");
function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
var OptionsPopup = /*#__PURE__*/ _react.default.forwardRef(function(props, ref) {
    var content_ref = _react.default.useRef(null);
    var options_ref = _react.default.useRef(null);
    var ref1 = _react.default.useState(false), show = ref1[0], setShow = ref1[1];
    var ref2 = _react.default.useState(false), eventAdded = ref2[0], setEventAdded = ref2[1];
    _react.default.useImperativeHandle(ref, function() {
        return {
            setInitPosition: function(element, options) {
                return (0, _jsManagement).setInitPosition(element, options_ref.current, options);
            },
            show: function() {
                setShow(true);
            },
            close: function() {
                setShow(false);
            }
        };
    });
    _react.default.useEffect(function() {
        if (!eventAdded) {
            (0, _jsManagement).OnOutsideClickListener(options_ref.current, function() {
                return setShow(false);
            });
            setEventAdded(true);
        }
    }, []);
    return(/*#__PURE__*/ _react.default.createElement("div", {
        ref: options_ref,
        className: "options__popup " + (show ? 'show' : '') + " " + (props.className ? props.className : "")
    }, props.children));
});
exports.OptionsPopup = OptionsPopup;
