// mirajs/dist/react/sidemenu.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.default = exports.SidemenuHeader = exports.SidemenuFooter = exports.SidemenuContainer = exports.SidemenuBanner = exports.Sidemenu = void 0;
var _react = _interopRequireDefault(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _js = require("../js");
var _index = require("./index");
function _assertThisInitialized(self) {
    if (self === void 0) {
        throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
    }
    return self;
}
function _classCallCheck(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
        throw new TypeError("Cannot call a class as a function");
    }
}
function _defineProperties(target, props) {
    for(var i = 0; i < props.length; i++){
        var descriptor = props[i];
        descriptor.enumerable = descriptor.enumerable || false;
        descriptor.configurable = true;
        if ("value" in descriptor) descriptor.writable = true;
        Object.defineProperty(target, descriptor.key, descriptor);
    }
}
function _createClass(Constructor, protoProps, staticProps) {
    if (protoProps) _defineProperties(Constructor.prototype, protoProps);
    if (staticProps) _defineProperties(Constructor, staticProps);
    return Constructor;
}
function _defineProperty(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}
function _extends() {
    _extends = Object.assign || function(target) {
        for(var i = 1; i < arguments.length; i++){
            var source = arguments[i];
            for(var key in source){
                if (Object.prototype.hasOwnProperty.call(source, key)) {
                    target[key] = source[key];
                }
            }
        }
        return target;
    };
    return _extends.apply(this, arguments);
}
function _getPrototypeOf(o) {
    _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) {
        return o.__proto__ || Object.getPrototypeOf(o);
    };
    return _getPrototypeOf(o);
}
function _inherits(subClass, superClass) {
    if (typeof superClass !== "function" && superClass !== null) {
        throw new TypeError("Super expression must either be null or a function");
    }
    subClass.prototype = Object.create(superClass && superClass.prototype, {
        constructor: {
            value: subClass,
            writable: true,
            configurable: true
        }
    });
    if (superClass) _setPrototypeOf(subClass, superClass);
}
function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _possibleConstructorReturn(self, call) {
    if (call && (_typeof(call) === "object" || typeof call === "function")) {
        return call;
    }
    return _assertThisInitialized(self);
}
function _setPrototypeOf(o, p) {
    _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) {
        o.__proto__ = p;
        return o;
    };
    return _setPrototypeOf(o, p);
}
var _typeof = function(obj) {
    "@swc/helpers - typeof";
    return obj && typeof Symbol !== "undefined" && obj.constructor === Symbol ? "symbol" : typeof obj;
};
function _isNativeReflectConstruct() {
    if (typeof Reflect === "undefined" || !Reflect.construct) return false;
    if (Reflect.construct.sham) return false;
    if (typeof Proxy === "function") return true;
    try {
        Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function() {}));
        return true;
    } catch (e) {
        return false;
    }
}
function _createSuper(Derived) {
    var hasNativeReflectConstruct = _isNativeReflectConstruct();
    return function _createSuperInternal() {
        var Super = _getPrototypeOf(Derived), result;
        if (hasNativeReflectConstruct) {
            var NewTarget = _getPrototypeOf(this).constructor;
            result = Reflect.construct(Super, arguments, NewTarget);
        } else {
            result = Super.apply(this, arguments);
        }
        return _possibleConstructorReturn(this, result);
    };
}
var SidemenuHeader = function(props) {
    var children = props.children, className = props.className;
    return(/*#__PURE__*/ _react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/ _react.default.createElement("div", _extends({}, props, {
        className: "dcx-sidemenu__header " + className
    }), children), /*#__PURE__*/ _react.default.createElement(_index.Divider, null)));
};
var SidemenuFooter = function(props) {
    var children = props.children, className = props.className;
    return(/*#__PURE__*/ _react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/ _react.default.createElement("div", _extends({}, props, {
        className: "dcx-sidemenu__footer " + className
    }), children)));
};
var SidemenuBanner = function(props) {
    var children = props.children, className = props.className;
    return(/*#__PURE__*/ _react.default.createElement("div", _extends({}, props, {
        className: "dcx-sidemenu__banner " + className
    }), children));
};
var SidemenuContainer = function(props) {
    var children = props.children, className = props.className;
    return(/*#__PURE__*/ _react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/ _react.default.createElement("div", _extends({}, props, {
        className: "dcx-sidemenu__container " + className
    }), children), /*#__PURE__*/ _react.default.createElement(_index.Divider, null)));
};
var Sidemenu = /*#__PURE__*/ function(_Component) {
    "use strict";
    _inherits(Sidemenu, _Component);
    var _super = _createSuper(Sidemenu);
    function Sidemenu(props) {
        _classCallCheck(this, Sidemenu);
        var _this;
        _this = _super.call(this, props);
        _defineProperty(_assertThisInitialized(_this), "sidemenuRef", undefined);
        _this.state = {
            open: false,
            id: props.id,
            width: props.width !== undefined ? props.width : 280,
            time: props.time !== undefined ? props.time : 200
        };
        return _this;
    }
    _createClass(Sidemenu, [
        {
            key: "componentDidMount",
            value: function componentDidMount() {
                this.sidemenuRef = (0, _js).DCSidemenu("#" + this.state.id, {
                    width: this.state.width,
                    time: this.state.time
                });
                console.log("Sidemenu Ref : ", this.sidemenuRef);
            }
        },
        {
            /**
     *
     * @param _param {{width: number, time: number}}
     */ key: "reinit",
            value: function reinit(_param) {
                this.sidemenuRef = (0, _js).DCSidemenu("#" + this.state.id, {
                    width: _param.width,
                    time: _param.time
                });
                this.setState({
                    width: _param.width,
                    time: _param.time
                });
                this.open();
            }
        },
        {
            key: "open",
            value: function open() {
                if (this.sidemenuRef !== undefined) this.sidemenuRef.open();
            }
        },
        {
            key: "close",
            value: function close() {
                if (this.sidemenuRef !== undefined) this.sidemenuRef.close();
            }
        },
        {
            key: "render",
            value: function render() {
                var _props = this.props, children = _props.children, id = _props.id, className = _props.className;
                return(/*#__PURE__*/ _react.default.createElement("div", _extends({}, this.props, {
                    className: "dcx-sidemenu " + className,
                    id: id
                }), children));
            }
        }
    ]);
    return Sidemenu;
}(_react.default.Component);
Sidemenu.propTypes = {
    id: _propTypes.default.string,
    time: _propTypes.default.number,
    width: _propTypes.default.number
};
exports.Sidemenu = Sidemenu;
exports.SidemenuBanner = SidemenuBanner;
exports.SidemenuContainer = SidemenuContainer;
exports.SidemenuFooter = SidemenuFooter;
exports.SidemenuHeader = SidemenuHeader;
var _default = Sidemenu;
exports.default = _default;
